formatter/src/main/java/org/togetherjava/formatter/Formatter.java

```diff
@@ -0,0 +1,319 @@
+package org.togetherjava.formatter;
+
+import org.togetherjava.formatter.tokenizer.Lexer;
+import org.togetherjava.formatter.tokenizer.Token;
+import org.togetherjava.formatter.tokenizer.TokenType;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Formatter which can format a given string into a string which contains code blocks etc
+ *
+ * @author illuminator3
+ */
+public class Formatter {
+    /**
+     * List of tokens who should not be put after a space
+     */
+    private static final List<TokenType> NON_SPACE_TOKENS = List.of(TokenType.DOT,
+            TokenType.SEMICOLON, TokenType.NOT, TokenType.OPEN_PARENTHESIS,
+            TokenType.CLOSE_PARENTHESIS, TokenType.OPEN_BRACKETS, TokenType.CLOSE_BRACKETS,
+            TokenType.SMALLER, TokenType.BIGGER, TokenType.COMMA, TokenType.FOR, TokenType.IF,
+            TokenType.WHILE, TokenType.PLUSPLUS, TokenType.MINUSMINUS, TokenType.RETURN,
+            TokenType.THIS, TokenType.PUBLIC, TokenType.PROTECTED, TokenType.PRIVATE, TokenType.TRY,
+            TokenType.CATCH, TokenType.PACKAGE, TokenType.METHOD_REFERENCE, TokenType.ELSE_IF,
+            TokenType.ELSE);
+
+    /**
+     * Formats the given tokens
+     *
+     * @param tokens tokens to format
+     * @return resulting code
+     * @author illuminator3
+     */
+    public String format(List<Token> tokens) {
+        List<Section> sections = sectionize(indexTokens(tokens));
+        StringBuilder result = new StringBuilder();
+
+        for (Section section : sections) {
+            if (section.isCodeSection()) {
+                result.append("```java\n")
+                      .append(writeCodeSection(section.tokens()))
+                      .append("\n```");
+            } else {
+                result.append(revert(section.tokens()));
+            }
+        }
+
+        return result.toString();
+    }
+
+    /**
+     * Formats the given string using a given lexer
+     *
+     * @param input input to format
+     * @param lexer lexer to use
+     * @return resulting code
+     * @author illuminator3
+     */
+    public String format(String input, Lexer lexer) {
+        return format(lexer.tokenize(input));
+    }
+
+    /**
+     * Reverts the tokens into the "original state" before the tokenization occurred
+     *
+     * @param tokens tokens to revert
+     * @return original state
+     * @author illuminator3
+     */
+    private String revert(List<Token> tokens) {
+        return tokens.stream().map(Token::content).collect(Collectors.joining());
+    }
+
+    /**
+     * Writes and formats a given code section (in form of a list of tokens) into a StringBuilder
+     *
+     * @param tokens tokens to write
+     * @return written code sections
+     * @author illuminator3
+     */
+    private StringBuilder writeCodeSection(List<Token> tokens) {
+        purgeWhitespaces(tokens = makeMutable(tokens));
```

Comment:
This is quite hard to read, could you split this into two lines and make a new variable for it? Ideally purgeWhitspace would return a new list, so that would be:
```java
List<Token> normalizedTokens = purgeWhitespaces(tokens);
```